File: sources/cpp/01_branches.cpp
// task 01 branches — expected output: 33333334 13333333 7619048 45714285
// build: g++ -O2 -pthread -o prog 01_branches.cpp    run: ./prog
// also builds with: clang++ -O2 -pthread -o prog 01_branches.cpp | cl /O2 /EHsc /Fe:prog 01_branches.cpp

#include <cstdio>

int main() {
    long long a = 0;
    long long b = 0;
    long long c = 0;
    long long d = 0;

    for (long long i = 0; i < 100000000LL; ++i) {
        if (i % 3 == 0) {
            a += 1;
        } else if (i % 5 == 0) {
            b += 1;
        } else if (i % 7 == 0) {
            c += 1;
        } else {
            d += 1;
        }
    }

    std::printf("%lld %lld %lld %lld\n", a, b, c, d);
    return 0;
}

File: sources/r/01_branches.R
# task 01 branches — expected output: 33333334 13333333 7619048 45714285
# build: Rscript 01_branches.R    run: Rscript 01_branches.R
#
# while, not for: for (i in 0:99999999) materialises a 400 MB integer vector
# before the first iteration. Counters are doubles; R has no 64-bit integers
# and every value here is exact in a double.

branches <- function() {
  a <- 0
  b <- 0
  c <- 0
  d <- 0
  i <- 0
  while (i < 100000000) {
    if (i %% 3 == 0) {
      a <- a + 1
    } else if (i %% 5 == 0) {
      b <- b + 1
    } else if (i %% 7 == 0) {
      c <- c + 1
    } else {
      d <- d + 1
    }
    i <- i + 1
  }
  cat(sprintf("%.0f %.0f %.0f %.0f\n", a, b, c, d))
}

branches()
